java/com/walmartlabs/logback/RiemannAppender.java

```diff
@@ -140,26 +150,47 @@ protected synchronized void append(E event) {
 
     if (isMinimumLevel(logEvent)) {
       EventDSL rEvent = createRiemannEvent(logEvent);
+
+      if (debug) {
+        printError("%s.append: sending riemann event: %s", className, rEvent);
+      }
+
       try {
-        try {
-          if (debug) {
-            printError("%s.append: sending riemann event: %s", className, rEvent);
-          }
-          rEvent.send();
-          if (debug) {
-            printError("%s.append(logEvent): sent to riemann %s:%s", className, riemannHostname, riemannPort);
-          }
-        } catch (Exception ex) {
-          if (debug) {
-            printError("%s: Error sending event %s", this, ex);
-            ex.printStackTrace(System.err);
-          }
 
-          riemannClient.reconnect();
-          rEvent.send();
-        }
-      } catch (Exception ex) {
-        // do nothing
+        int attempt = 0;
+        while (true) {
+          try {
+            if (++attempt > 1) {
+              riemannClient.reconnect();
+            }
+
+            Proto.Msg ack = rEvent.send().deref();
+
+            if (!ack.getOk()) {
+              throw new IOException("Error from Riemann server: " + ack.getError());
+            }
+
+            if (debug) {
+              printError("%s.append(logEvent): sent to riemann %s:%s", className, riemannHostname, riemannPort);
+            }
+
+            // Ok to break out of the retry loop once successful.
+            break;
+
+          } catch (Exception ex) {
+            if (debug) {
+              printError("%s: Error sending event %s", this, ex);
+            }
+
+            if (attempt < MAX_ATTEMPTS) {
+              continue;
+            }
+
+            throw new IOException(ex.getMessage(), ex);
+          }
+        } // retry loop
+      } catch (IOException ex) {
+        // Failure after retries exhausted:
         if (debug) {
           printError("%s.append: Error during append(): %s", className, ex);
```

Comment:
It might it make sense to pop this line out of the `debug` check. At it stands, we're totally blind as to whether events are occasionally failing unless we turn `debug` on.